File: tests/assistant-response-stream.test.mjs
import { describe, expect, it } from 'vitest';
import { readFileSync } from 'node:fs';
import {
  decodeAssistantResponseFrame,
  encodeAssistantResponseFrame,
} from '../src/contracts/assistant-response-stream.mjs';

const RESPONSE_ID = '01ARZ3NDEKTSV4RRFFQ69G5FAV';
const SOURCE_EVENT_ID = '01ARZ3NDEKTSV4RRFFQ69G5FAW';
const vectors = JSON.parse(readFileSync(new URL('./fixtures/protocol/mina-chat-payload-v2-vectors.json', import.meta.url), 'utf8')).vectors;

describe('contrat des réponses streamées Mina', () => {
  it('conserve le même lien source et des séquences strictes du début à la fin', () => {
    const started = decodeAssistantResponseFrame(encodeAssistantResponseFrame({
      type: 'assistant.response.started',
      responseId: RESPONSE_ID,
      sourceEventId: SOURCE_EVENT_ID,
      sequence: 0,
    }));
    const chunk = decodeAssistantResponseFrame(encodeAssistantResponseFrame({
      type: 'assistant.response.chunk',
      responseId: RESPONSE_ID,
      sourceEventId: SOURCE_EVENT_ID,
      sequence: 1,
      text: 'bon',
    }));
    const completed = decodeAssistantResponseFrame(encodeAssistantResponseFrame({
      type: 'assistant.response.completed',
      responseId: RESPONSE_ID,
      sourceEventId: SOURCE_EVENT_ID,
      sequence: 2,
      text: 'bonjour',
    }));

    expect(started).toEqual({
      type: 'assistant.response.started', responseId: RESPONSE_ID, sourceEventId: SOURCE_EVENT_ID,
      sequence: 0, text: null, code: null,
    });
    expect(chunk).toMatchObject({ type: 'assistant.response.chunk', sequence: 1, text: 'bon' });
    expect(completed).toMatchObject({ type: 'assistant.response.completed', sequence: 2, text: 'bonjour' });
  });

  it('refuse un fragment ambigu, un mauvais ordre initial ou un identifiant non ULID', () => {
    expect(() => encodeAssistantResponseFrame({
      type: 'assistant.response.started', responseId: RESPONSE_ID, sourceEventId: SOURCE_EVENT_ID,
      sequence: 0, text: 'ne doit pas exister',
    })).toThrow('assistant_response_started_body_invalid');
    expect(() => encodeAssistantResponseFrame({
      type: 'assistant.response.chunk', responseId: RESPONSE_ID, sourceEventId: SOURCE_EVENT_ID,
      sequence: 0, text: 'bon',
    })).toThrow('assistant_response_sequence_invalid');
    expect(() => encodeAssistantResponseFrame({
      type: 'assistant.response.completed', responseId: 'not-ulid', sourceEventId: SOURCE_EVENT_ID,
      sequence: 2, text: 'bonjour',
    })).toThrow('assistant_response_id_invalid');
  });

  it('rejoue le vecteur de réponse finale produit par Node et borne les codes d’échec', () => {
    const vector = vectors.find(({ name }) => name === 'assistant_response_completed');
    const payload = encodeAssistantResponseFrame({
      type: vector.type,
      responseId: vector.meta.responseId,
      sourceEventId: vector.meta.sourceEventId,
      sequence: vector.meta.sequence,
      text: Buffer.from(vector.binaryHex, 'hex').toString('utf8'),
    });

    expect(payload.toString('hex')).toBe(vector.payloadHex);
    expect(decodeAssistantResponseFrame(payload)).toMatchObject({
      type: 'assistant.response.completed', sequence: 2, text: 'bonjour', code: null,
    });
    expect(decodeAssistantResponseFrame(encodeAssistantResponseFrame({
      type: 'assistant.response.failed', responseId: RESPONSE_ID, sourceEventId: SOURCE_EVENT_ID,
      sequence: 3, code: 'provider_timeout',
    }))).toMatchObject({ type: 'assistant.response.failed', sequence: 3, text: null, code: 'provider_timeout' });
    expect(() => encodeAssistantResponseFrame({
      type: 'assistant.response.failed', responseId: RESPONSE_ID, sourceEventId: SOURCE_EVENT_ID,
      sequence: 3, code: 'arbitrary_provider_detail',
    })).toThrow('assistant_response_code_invalid');
  });
});

File: android/core/protocol/src/test/kotlin/fr/mina/gateway/protocol/AssistantResponseStreamTest.kt
package fr.mina.gateway.protocol

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Test

class AssistantResponseStreamTest {
    private companion object {
        const val RESPONSE_ID = "01ARZ3NDEKTSV4RRFFQ69G5FAV"
        const val SOURCE_EVENT_ID = "01ARZ3NDEKTSV4RRFFQ69G5FAW"
    }

    @Test
    fun `conserve la correlation et les sequences du debut a la fin`() {
        val started = AssistantResponseStream.decode(
            ChatPayloadCodec.decode(AssistantResponseStream.encodeStarted(RESPONSE_ID, SOURCE_EVENT_ID))
                as ChatPayloadCodec.PayloadV2,
        )
        val chunk = AssistantResponseStream.decode(
            ChatPayloadCodec.decode(AssistantResponseStream.encodeChunk(RESPONSE_ID, SOURCE_EVENT_ID, 1, "bon"))
                as ChatPayloadCodec.PayloadV2,
        )
        val completed = AssistantResponseStream.decode(
            ChatPayloadCodec.decode(AssistantResponseStream.encodeCompleted(RESPONSE_ID, SOURCE_EVENT_ID, 2, "bonjour"))
                as ChatPayloadCodec.PayloadV2,
        )

        assertEquals("assistant.response.started", started.type)
        assertEquals(0, started.sequence)
        assertNull(started.text)
        assertEquals("assistant.response.chunk", chunk.type)
        assertEquals("bon", chunk.text)
        assertEquals("assistant.response.completed", completed.type)
        assertEquals("bonjour", completed.text)
    }

    @Test
    fun `refuse un fragment ambigu une sequence initiale invalide et un identifiant invalide`() {
        assertThrows(IllegalArgumentException::class.java) {
            AssistantResponseStream.encodeStarted(RESPONSE_ID, SOURCE_EVENT_ID, "ne doit pas exister")
        }
        assertThrows(IllegalArgumentException::class.java) {
            AssistantResponseStream.encodeChunk(RESPONSE_ID, SOURCE_EVENT_ID, 0, "bon")
        }
        assertThrows(IllegalArgumentException::class.java) {
            AssistantResponseStream.encodeCompleted("not-ulid", SOURCE_EVENT_ID, 2, "bonjour")
        }
    }

    @Test
    fun `encode un echec borne sans corps en clair`() {
        val failed = AssistantResponseStream.decode(
            ChatPayloadCodec.decode(AssistantResponseStream.encodeFailed(RESPONSE_ID, SOURCE_EVENT_ID, 3, "provider_timeout"))
                as ChatPayloadCodec.PayloadV2,
        )

        assertEquals("assistant.response.failed", failed.type)
        assertEquals(3, failed.sequence)
        assertNull(failed.text)
        assertEquals("provider_timeout", failed.code)
        assertThrows(IllegalArgumentException::class.java) {
            AssistantResponseStream.encodeFailed(RESPONSE_ID, SOURCE_EVENT_ID, 3, "provider_detail_unbounded")
        }
    }
}
